File: pset6/credit.py
get_input = True
while get_input:
    try:
        credit = int(input("Number: "))
    except ValueError:
        pass
    else:
        if credit > 0:
            get_input = False

def checksum(card):
    reverse = card [::-1]
    sum = 0
    for i in range(1, len(reverse), 2):
        temp = str(int(reverse[i]) * 2)
        if (int(temp) < 10):
            sum += int(temp)
        else:
            sum += (int(temp[0]) + int(temp[1]))
    for i in range(0, len(reverse), 2):
        sum += int(reverse[i])
    if sum % 10 == 0:
        return True
    else:
        print("INVALID")
        return False

if (checksum(str(credit))):
    card = str(credit)
    if (card[0] == "3" and (card[1] in ["4", "7"])):
        print("AMEX")
    elif (card[0] == "5" and (card[1] in ["1", "2", "3", "4", "5"])):
        print("MASTERCARD")
    elif (card[0] == "4"):
        print("VISA")
    else:
        print("INVALID")

File: pset1/credit.c
#include <stdio.h>

int main (void)
{
	// paypal dummy cards https://developer.paypal.com/docs/payflow/payflow-pro/payflow-pro-testing/#credit-card-numbers-for-testing
	
	long int creditcard;
	int cardarray[16];
	int sum = 0;
	char type = '\0'; // m is MASTERCARD, v is VISA, a is AMEX, o is OTHER
	
	printf("Credit Card Number: ");
	scanf("%li", &creditcard);
	
	// init array
	long int duplicate = creditcard;
	for (int i = 0; i < 16; i++)
	{
		cardarray[i] = duplicate % 10;
		duplicate /= 10;
	}
	
	int counter = 0;
	duplicate = creditcard;
	while (duplicate != 0)
	{
		counter++;
		duplicate /= 10;
	}
	
	duplicate = creditcard;
	while (duplicate != 0) {
		if ((duplicate == 34 || duplicate == 37) && counter == 15)
		{
			type = 'a'; // amex
		}
		else if (((duplicate >= 2221 && duplicate <= 2720) || (duplicate >= 50 && duplicate <= 69)) && counter == 16)
		{
			type = 'm'; // mastercard
		}
		else if (duplicate >= 40 && duplicate <= 49 && (counter == 13 || counter == 16))
		{
			type = 'v'; // visa
		}
		else if (type == '\0')
		{
			type = 'o'; // other
		}
		duplicate /= 10;
	}
	
	// checksum
	for (int i = 1; i < 16; i+=2)
	{
		cardarray[i] *= 2;
	}
	for (int i = 0; i < 16; i++)
	{
		if (cardarray[i] < 10)
		{
			sum += cardarray[i];
		}
		else
		{
			sum += (cardarray[i] % 10);
			sum += cardarray[i]/10;
		}
	}
	
	if (sum % 10 == 0)
	{
		printf("VALID\nCARD TYPE : ");
		if (type == 'a')
		{
			printf("AMEX\n");
		}
		else if (type == 'm')
		{
			printf("MASTERCARD/MAESTRO\n");
		}
		else if (type == 'v')
		{
			printf("VISA\n");
		} 
		else if (type == 'o')
		{
			printf("OTHER\n");
		}
		else
		{
			printf("program error please call 911.\n");
		}
	} else
	{
		printf("INVALID\n");
	}
	
	return 0;
}